### ISSUE TITLE
Multiple Assertions

### ISSUE THREAD

**Issue by circa1741** [nunit/nunit#391]
May I ask if there is a plan to have something similar to MbUnit's Multiple Assertions capability?

This might provide some insight:
https://code.google.com/p/mb-unit/source/browse/trunk/v3/src/MbUnit/MbUnit/Framework/Assert.cs?r=1570#282

Evaluates a block of code that contains multiple related assertions.

While the action delegate runs, the behavior of assertions is change such that
failures are captured but do not cause a <see cref="AssertionFailureException" />
to be throw.  When the delegate returns, the previous assertion failure behavior
is restored and any captured assertion failures are reported.  The net effect
of this change is that the test can continue to run even after an assertion failure
occurs which can help to provide more information about the problem.

A multiple assertion block is useful for verifying the state of a single
component with many parts that require several assertions to check.
This feature can accelerate debugging because more diagnostic information
become available at once.

Also:
https://code.google.com/p/mb-unit/source/browse/trunk/v3/src/MbUnit/MbUnit/Framework/MultipleAssertsAttribute.cs?r=1570

Runs the test as if it were surrounded by <see cref="Assert.Multiple(Action)" /> so that
multiple assertion failures within the test are tolerated.

When an assertion failure occurs, it is reported but the test is allowed to proceed
until it completes or throws an unhandled exception.  When the test finishes, it will
still be marked as failed, as usual.
## 

Partial specs as entered below by @CharliePoole 

There are a number of things needed here and I guess it's time to make a list! I may end up breaking some of these out into separate issues, but for the moment let's just stick with bullet points.

- We will have to redesign the xml result format to support multiple results per test case. Currently, each test can only have a single result and tests are nested inside other tests (suites). However, if each assertion can generate a separate result, we need to hold them somewhere. This might be, for example, a `<results>` element containing one or more `<result>` elements and superseding some existing elements.
- The internal framework TestResult Type will need to hold multiple results and generate xml as designed in the first bullet.
- Clients and the engine will need a way to recognize which format is being used.
- Clients and the engine may need a way to control which format gets used.
- We'll have to firm up the syntax we intend to support. This is the part we have mostly discussed here, so I think it's pretty ready.
- We need to introduce a mechanism for reporting results during a test without generating an exception, which would, of course, stop the test.
- Finally, we have to implement the syntax for multiple assertions.
 

The preliminary part of this, supporting multiple results per test, will actually be useful for three different features we would like to have:
1. Multiple Assertions (this issue)
2. Warning Assertions - because allowing continuation after a warning implies multiple results
3. Reporting passed assertions.

If anyone wants to pull out a part of this and work on it, let me know.

**Comment by svaze** [comment]
Hey Charlie,
This feature is one of the best features for functional automated testing.
Many users out there will love to have this feature. Personally my favour for mbUnit was for the following reasons,
1) MultipleAssert support, 
2) Parallelism and 
3) Screenshot appending/attaching capabilities to Test Log.
These all features were not present in NUnit. However, mbUnit has now stopped any development and provides virtually no support now, so its becoming obsolete to work with.
With NUnit 3.0, parallelism is available which will make it popular for Functional automated testing community if along with it, features like MultipleAsserts and Screenshot appending are included in it.

Many thanks for such a wonderful product.

Thanks, Shrirang Vaze

**Comment by CharliePoole** [comment]
@svaze Thanks for the input. As you point out, this feature is particularly useful for functional testing. We have kept it in the plan for the 3.0 release for exactly that reason. OTOH, since it's less useful for unit testing, which has always been the core of NUnit, we have not yet given it a high priority.

The status is pending because of the internal changes needed to implement it, as described in some of the comments above. Allowing a Test to have multiple Results is a fairly big design change and potentially disruptive to the rest of NUnit. We will continue to work on this idea and try to keep it in the 3.0 release, but it's possible it will get pushed to the following release, depending on our resources. It's more important to us to actually _have_ a 3.0 release!

The feature involving attachments has never been requested. If you would like to create an issue for it, that would be a welcome contribution. You should provide fairly complete detail on how it would work, how it would be used by tests, etc. Note that saying it should work "like mbUnit" won't really do the job for us. Only a few folks working on the project are familiar with mbUnit. If only those few can understand your request, then it's much less likely to be implemented. Fair enough?

**Comment by shortstacked** [comment]
Hey,
I have a question about this feature. I'm using NUnit.3.0.0-rc-2, and I am able to see the Assert.Multiple() method which accepts a TestDelegate, however when using this with multiple asserts the first failure causes it to break out and report. 

I can see the milestone was modified but I'm unsure if that means available as a whole or only partially available as the [MultipleAssert] attribute also does not seem to be available.

Would you be able to advise if this feature is available or if the multiple assert feature is just not fully implemented yet.

Thanks.

**Comment by CharliePoole** [comment]
As indicated on #391, the feature has been postponed until NUnit 3.2.

Assert.Multiple is a stub method which seems to have been committed
accidentally. All it does is execute the delegate you give it.

**Comment by SuperDrew** [comment]
@CharliePoole This is still blocked, any idea when it may become unblocked?  

This would be extremely useful in the stochastic testing methods I employ.  I would like to group a potentially very large set of random test cases.  I also want all the test cases to run even if some fail. This looks like a great solution for this problem.  I can give an example of what my tests look like if you're interested. The alternative has been to build my own test result and report generation which seems like a waste of effort as it essentially re-writes some of what is in NUnit already.

**Comment by rprouse** [comment]
@SuperDrew, this is one of those issues that we would like to do, but still haven't found a clean way to do it within the framework as it currently stands. Thus we can't give you an estimate.

**Comment by novaterata** [comment]
@SuperDrew In the meantime, you can write your own CustomConstraints that execute all the assertions and aggregate the expected and results in the messages. I recently had to do this in JUnit (also had to implement a non-short-circuit allOf) and I'm quite happy with the results. It might take a little longer to write, but well worth it.

**Comment by SuperDrew** [comment]
@rprouse No problem, thanks for the info. And thank you for an awesome tool!
@novaterata I don't think that'll work in NUnit, or at least I wouldn't know how to prevent an Assertion failure from stopping the execution.

**Comment by novaterata** [comment]
Right, you would write it as a single assertion using boolean logic.

**Comment by CharliePoole** [comment]
I've left this as blocked and also assigned to myself so that nobody would start down an untenable path to implement a half-*ssed solution. Individuals often write their own workarounds as @novaterata suggested, accumulating errors and then finally making an assertion that there are no accumulated errors. It's actually possible to do it that way inside NUnit and I even wrote some throw-away code to do it. But in the end, that code was __really__ only worth throwing away. :-(

There are a number of things needed here and I guess it's time to make a list! I may end up breaking some of these out into separate issues, but for the moment let's just stick with bullet points.

* We will have to redesign the xml result format to support multiple results per test case. Currently, each test can only have a single result and tests are nested inside other tests (suites). However, if each assertion can generate a separate result, we need to hold them somewhere. This might be, for example, a `<results>` element containing one or more `<result>` elements and superseding some existing elements.

* The internal framework TestResult Type will need to hold multiple results and generate xml as designed in the first bullet. 

* Clients and the engine will need a way to recognize which format is being used.

* Clients and the engine **may** need a way to control which format gets used.

* We'll have to firm up the syntax we intend to support. This is the part we have mostly discussed here, so I think it's pretty ready.

* We need to introduce a mechanism for reporting results during a test without generating an exception, which would, of course, stop the test.

* Finally, we have to implement the syntax for multiple assertions.

The preliminary part of this, supporting multiple results per test, will actually be useful for three different features we would like to have:

1. Multiple Assertions (this issue)

2. Warning Assertions - because allowing continuation after a warning implies multiple results

3. Reporting passed assertions.

If anyone wants to pull out a part of this and work on it, let me know.

**Comment by jcansdale** [comment]
How about something like this:

```
        // HACK: Just to demo functionality. ;)
        [TestCase(0), TestCase(1), TestCase(2)]
        public void TestAction(int i)
        {
            var actions = new List<Action>(MultiAsserts());
            actions[i]();
        }

        // [TestActions] attribute?
        public IEnumerable<Action> MultiAsserts()
        {
            var hello = "hello, world";
            yield return () => Assert.That(hello, Does.StartWith("H"));
            yield return () => Assert.That(hello, Does.EndWith("!"));
            yield return () => Assert.That(hello, Does.Contain("World"));
        }
```

**Comment by jcansdale** [comment]
You could even get fancy and pull out a test name from the assert expression:

```
        [TestCaseSource(nameof(MultiAsserts))]
        public void TestAction(Expression<Action> expr)
        {
            TestContext.WriteLine("TestName: " + expr);
            expr.Compile().Invoke();
        }

        static public IEnumerable<Expression<Action>> MultiAsserts()
        {
            var hello = "hello, world";
            yield return () => Assert.That(hello, Does.StartWith("H"));
            yield return () => Assert.That(hello, Does.EndWith("!"));
            yield return () => Assert.That(hello, Does.Contain("World"));
        }
```

**Comment by novaterata** [comment]
Oh, just wanted to mention another workaround that is much quicker than
CustomConstraint. Simply populate expected and actual anonymous objects
with the same properties that represent the things you want to assert and
Assert.That(actual, Is.EqualTo(expected). Since anonymous objects behave
like value types it will work, and the failure message should be detailed
enough.

**Comment by rprouse** [comment]
@CharliePoole I have taken your bullet list and copied it to the top of the issue so that it does not get lost in the middle and we don't need to search through all the comments. For now, I changed it into a checklist, but we might want to make this an epic.

>  We'll have to firm up the syntax we intend to support. This is the part we have mostly discussed here, so I think it's pretty ready.

For the syntax that this should eventually take, I like the MbUnit format outlined by @circa1741 near the start of this issue. I think it is simple, clean and readable.

**Comment by CharliePoole** [comment]
@rprouse I had no idea you could do a checklist!

I was planning to let the list of tasks percolate for a bit and then create issues where appropriate. It's slightly tricky to subdivide into issues where each one is independently mergeable, but I think there's a path. Let's see if anyone else comments in the next few days, since there has been a lot of interest in this.

**Comment by CharliePoole** [comment]
@novaterata Yes, what you are describing is sort of the "standard" approach where you accumulate results in the test itself and create a composite result.  I did a spike where NUnit itself did the accumulating and produced a composite message, but it wasn't very satisfactory.

@jcansdale This seems much more complicated for users than the mbunit syntax I incorporated in the spec - https://github.com/nunit/docs/wiki/Multiple-Asserts-Spec

**Comment by jcansdale** [comment]
I wondered if it was something I could hack up using custom attributes. Where is the best place to find information about how to use them? There only seems to be a stub on the wiki at the moment:
https://github.com/nunit/docs/wiki/Custom-Attributes

**Comment by rprouse** [comment]
All of the existing NUnit attributes are written the same as you would a
custom attribute, so the code is the best example.

**Comment by CharliePoole** [comment]
@jcansdale I don't believe you can. As you see, it requires changes in the framework, the format of the XML, the engine and the console runner. The gui will need to reflect it as well. Allowing a test to have multiple results is a significant change.

All the hacks people do try to simulate this by having one result with values aggregated. The messages and stack traces need to be somehow combined to trick NUnit into treating it as one result. Any runners wanting to display the info would have to parse the pseudo-result into it's component parts. I don't think we want to spend time doing that, especially since people can already do it in their test code as a workaround.

In any case, WRT the user syntax I feel we have __almost__ selected one. If you are questioning whether it's a good idea it's important to resolve this because nobody will do any coding until we do.

**Comment by jcansdale** [comment]
I'm interested to see what's possible using the existing extensibility model before we add more complexity. I know from a test runner developer's POV, handling multiple results for a single test would be a pain. If we could instead handle it by exposing each assert as a test, there would be no changes at the test runner interface end.

I think the following could work reasonably well (note the 3, so test runners know how many tests/asserts to expect ahead of time):

```
        [MultipleAsserts(3)]
        public IEnumerable<Action> TestWitAsserts()
        {
            var hello = "hello, world";
            yield return () => Assert.That(hello, Does.StartWith("H"));
            yield return () => Assert.That(hello, Does.EndWith("!"));
            yield return () => Assert.That(hello, Does.Contain("World"));
        }
```

If you think the yield return syntax is too ugly, we could maybe do a similar thing like this:

```
        [MultipleAsserts(3)]
        public void TestWitAsserts()
        {
            Assert.Multiple( () =>
            {
                Assert.That(hello, Does.StartWith("H"));
                Assert.That(hello, Does.EndWith("!"));
                Assert.That(hello, Does.Contain("World"));
            }
        }
```

Where `Assert.Multiple` is a convenience method that is implemented like this:

```
public static void Multiple(Expression<Action>> asserts)
{
    // Extract Assert statements from `asserts` expression and pass them back to
    // [MultipleAsserts(...)] test implementation.
}
```

I wonder how Assert.Multiple is implemented in MbUnit?

**Comment by CharliePoole** [comment]
I'm cool with your thinking outside the box but let's remember that the user already has the capability to write this as separate tests. Folks asked for this because they actually want one test to have multiple assertions, all of  which are checked even if one fails. The syntax in the spec is something we have been promising for a while, so any replacement syntax has to be relatively easy to use.

"If we could instead handle it by exposing each assert as a test, there would be no changes at the test runner interface end."

Could you explain this more fully? It seems like this is what motivates your approach, but I don't get it. How would this syntactic approach change the other non-framework bullet points listed?

If you look at how it's implemented in MbUnit, please don't submit any code to NUnit on this issue. Our licensing is not compatible.

**Comment by jcansdale** [comment]
> Could you explain this more fully? It seems like this is what motivates your approach, but I don't get it. How would this syntactic approach change the other non-framework bullet points listed?

Current test runner interfaces know how to represent tests that pass or fail, but not a bag of assertion results. Implementing this in an intuitive way would  be a challenge.

If multiple assertions could instead be represented as individual tests, no changes would be needed to represent these results. I'm thinking about this for all test runners, not just TestDriven.Net. :wink: 

For example, this is how my suggestion would show up in NCrunch:

![ncrunch_multi_asserts](https://cloud.githubusercontent.com/assets/11719160/17616873/b926394e-606e-11e6-910e-d736b4a012a1.png)

It's pretty clear which assertions have failed and easy to find their message. I don't know how @remcomulder would handle this if a single test returned a bag of assertion results?

Another questions is, do we want the assertion results to be made available in real time as the asserts are executed (important for long running UI tests)? This is easy if each assert is surfaced as a separate test.

**Comment by remcomulder** [comment]
Just adding my 2 cents here.

I like the concept of being able to have multiple assertion results in a test.  It seems like it would be valuable.

However, James is not wrong in his concerns about the situation with test runners.  Although MbUnit does/did have a feature like this, it was fairly limited in its reporting capability and as far as I know, it didn't really require extra consideration by runners (at least, no one reported any bugs to me about it).

An approach like the one suggested would create new concerns that are unique to NUnit v3+.  NCrunch's data model was designed with some consideration for these concerns but the concept is so far completely untested.  It's safe to assume there will be quite a bit of code out there in the world that has been written in the expectation that each test can only return a single exception/assertion result.

I'm not saying that it isn't worthwhile, but rather that we don't really know what the risks are of implementing this ... and we won't know them until the feature is introduced and lots of people start encountering unexpected behaviour.

**Comment by CharliePoole** [comment]
@jcansdale @remcomulder 
We definitely have to consider the impact on runners (that was listed in my bullet points about this) and by that I don't just mean the NUnit runners but yours as well.

But here's the thing. This feature has had a syntax described for over a year. It wasn't moving forward because of considerations beyond the syntax. So I made this big bulleted list of the __other__ stuff that is holding us back. That includes the question of how runners will deal with it. Nevertheless, we are talking about this in terms of syntax.

In my view, the syntax the user uses - and certainly how we implement that syntax - is completely orthogonal to how runners deal with us. We should be talking about the information that flows back to the runner as XML (my first bullet) and how runners determine and/or control the particular format that gets used. We should be talking about how a runner gracefully degrades if it doesn't support the feature. The user syntax is where we started because a user created the issue. But it's not where we need to start for the design.

@jcansdale 's input is really valuable to me because I had not considered how runners that integrate with the source code would deal with this, only runners like the NUnit gui. We need to consider the special problems of such runners. But we shouldn't be doing it by restricting the syntax. We should trust that we can make any syntax look like anything we want to the runners.

@rprouse called my bullet points a "spec" but they are not. I **have** written a spec for the syntax part that's linked above, but the other bullets are simply vague pointers to areas for which we need specs. I'll write them when we have had enough discussion for me to do so. I listed them in the order in which I felt they had to be resolved. I could be wrong about the order, but I'm pretty sure that the details of the user syntax should not be the first thing we talk about.

**Comment by CharliePoole** [comment]
One difference between an Assertion treated as a "test" and a normal test is that you want to allow selected execution normal tests. Running an assert by itself shouldn't be possible.

**Comment by jcansdale** [comment]
I was thinking this would still be possible, but executing any of the assert tests would trigger execution of the original test method (up to that assert). I'm sure the devil will be in the details. :wink: 

I've added a new issue here #1737, so as not to clutter up this one. It might turn out to be impossible, but I'm interested to find out. :smile:

**Comment by CharliePoole** [comment]
Devil in the details for sure! As well as in the mind of the user who has to understand what's happening! I commented separately on #1737.

Now let's all get back to how this will look to runners!!!

**Comment by CharliePoole** [comment]
As you see, I made this into an Epic. (If you don't see, you have to install ZenHub.)

I think #1743 is the key item requiring some decisions in order to move ahead.

**Comment by yetibrain** [comment]
IMHO there is no need to use an extra Attribute in order to mark a test as a multi-assert test. It's unflexible. Why not allow common Asserts along with new classes like SingleAssert? SingleAssert objects could be added to a MultiAssert object. Like this, common assert can lead to assertions as usual while several single-asserts, added to a multi-assert object fulfills the multi-Assertion Feature.

**Comment by CharliePoole** [comment]
@yetibrain I agree that there's no need... but it's part of the syntax that somebody requested because it happens to match what they use in mbUnit. I see it as a marginally convenient shorthand that equates to wrapping all the code in an Assert.Multiple block.

When I made this issue into an Epic, I didn't create a separate issue for designing the syntax because I thought it had been settled for more than a year. It sounds like you would like to discuss an alternative syntax, as has also been proposed by @jcansdale in #1737. Do we need to make a new issue for the syntax? Right now, I'm focused on what the result file will look like (#1743) but I don't think more discussion of the syntax would hurt us and it's pretty much independent of the shape of the output.

**Comment by CharliePoole** [comment]
After a great deal of discussion under issue #1802, now closed, the spec on the wiki has been updated:
https://github.com/nunit/docs/wiki/Multiple-Asserts-Spec

Implementation will proceed under this Epic as soon as the other sub-issues are resolved.

**Comment by CharliePoole** [comment]
Cool! It turns out you can add issues from multiple repositories to an epic!

**Comment by CharliePoole** [comment]
I think we have one loose end in the implementation. I'll lay it out here and we can decide how to proceed, creating another sub-issue if necessary.

The spec describes actions to take on early termination. See https://github.com/nunit/docs/wiki/Multiple-Asserts-Spec#early-termination

In each case, if there are pending failures, we are supposed to give a special message. This is not yet implemented. It's not clear whether we should list the failures in some or all of these cases.

@nunit/core-team Should we implement messages as described? Should we list info about the pending failures?

**Comment by CharliePoole** [comment]
I'm going ahead and creating a separate issue, but I'm still looking for comments from @nunit/core-team and @nunit/contributors

**Comment by circa1741** [comment]
I vote for implementing the special message for pending failures and listing it for all of the cases.

**Comment by CharliePoole** [comment]
I created issue #1907 to handle this in nunitlite and nunit/nunit-console#136 for the console runner.

**Comment by CharliePoole** [comment]
Closing the epic itself, now that all the issues are complete.
